docs: idp-groups setup and migration (#2069)
canonical/jimm: docs/howto/manage-groups.md

>
> See also: {ref}`group`

```{important}
**Deprecation notice — migration to IDP groups**

> JAAS local groups (created with `juju add-group`) are deprecated and will be **removed** in the next major release of JAAS on the `v4` track. They are replaced by **IDP groups**, which are groups that your Identity Provider (IdP) already manages and sends to JAAS as an extra claim in the user's access token at login time.

Reviewer: ```suggestion
JAAS groups (created with `juju add-group`) are deprecated and will be **removed** in the next major release of JAAS on the `v4` track. They are replaced by **groups managed by the Identity Provider**.
```

> With IDP groups you no longer need to create groups in JAAS or assign users to them — user-to-group membership is owned by the IdP. In JAAS you only need to **assign permissions to an IDP group**, and every user who logs in with that group claim will inherit the permission automatically.

Reviewer: either user IDP or IdP (i prefer this one), but not both.

> With IDP groups you no longer need to create groups in JAAS or assign users to them — user-to-group membership is owned by the IdP. In JAAS you only need to **assign permissions to an IDP group**, and every user who logs in with that group claim will inherit the permission automatically.

Reviewer: ```suggestion
With Identity Provider (IdP) groups you no longer need to create groups in JAAS or assign users to them — group membership is managed by the IdP. In JAAS you only need to **assign permissions to an IdP group**, and every member of group inherits the permission automatically.
```

Reviewer 2: — could be ,

> Local groups remain fully supported on the current `v3` track. If you are starting a new deployment, prefer IDP groups from the outset. If you already use local groups, see {ref}`migrate-to-idp-groups` below for the migration path.

Reviewer: ```suggestion
JAAS groups remain fully supported on the current `v3` track. If you are starting a new deployment, prefer IdP groups from the outset. If you already use JAAS groups, see {ref}`migrate-to-idp-groups` below for the migration path.
```

```
````

> The sections below describe the **local groups** workflow, which is still supported on the `v3` track. For the IDP groups workflow and the migration from local groups to IDP groups, see {ref}`idp-groups` and {ref}`migrate-to-idp-groups`.

Reviewer: i would not talk about "local" group, but use "JAAS groups" instead.

```

> See more: {doc}`juju remove-group <../reference/jaas-plugin>`

(idp-groups)=
## About IDP groups

> See also: {ref}`group`

> An **IDP group** is a group that is owned and managed by your Identity Provider (IdP), not by JAAS. When a user logs in to JAAS, the IdP includes an extra claim in the access token that lists the groups the user belongs to. JAAS reads these claims and treats each value as an IDP group the user is a member of.

Reviewer: group being "owned" by and IdP is a bit weird.. "manged" is a better word

> Compared to local groups, the key difference is that **JAAS never creates IDP groups or assigns users to them** — that is entirely the IdP's responsibility. In JAAS you only assign permissions (e.g. `administrator` on a model) to an IDP group, and every user who logs in with that group claim inherits the permission automatically.

Reviewer 2: ```suggestion
Compared to JAAS groups, the key difference is that **JAAS no longer creates groups or assigns users to them**. In JAAS you only assign permissions (e.g. `administrator` on a model) to an IdP group, then every user who logs in with that group claim inherits the permission automatically.
```

(migrate-to-idp-groups)=
## Migrate from local groups to IDP groups

Local groups are deprecated and will be removed in the `v4` release of JAAS. To migrate, move group membership management to your IdP and replace local-group permission assignments with IDP-group permission assignments.

The high-level migration steps are:

1. **Configure your IdP** to emit a group claim (e.g. `groups`) that contains the group names you want to use in JAAS. The exact configuration depends on your IdP (e.g. Canonical Identity Platform / Keycloak, Auth0, etc.).
2. **Map the existing local groups** in JAAS to the group names the IdP will emit. For each local group, make sure the IdP emits a group with a matching name, and assign the same users to that group in the IdP.
3. **Re-assign permissions** from the local group to the IDP group. For every permission previously granted to `group-<name>#member`, grant the same permission to `idpgroup-<name>#member`.
4. **Remove the local group** permission assignments and the local group itself once the IDP group is in use and users have confirmed they inherit the expected access.

> ### Terraform migration example

Reviewer 3: I'm not sure this Terraform example belongs here. I don't mind it personally but could we link to the Terraform docs instead?

Reviewer 2: It's a bit difficult to find a better place for it, though.

It's also tricky to come up with better syntax for expressing the migration.

Reviewer 2: I do agree that it feels a bit weird. If we weren't also working on the provider and so familiar with it, we wouldn't even consider this.

Author: At the end of the day this seems public doc but it's indeed a guide for IS to migrate to idp group, and they specifically use the terraform provider. So it's useful to have an example here, on how to do it.

We can even remove this page at one point, since nobody will use the old jaas-groups

IdP groups are referenced using the `IdPgroup` tag and require the `#member` userset, just like JAAS groups. For example:

```text
# Grant administrator access on a model to every member of the "canonical" IdP group:
juju add-permission IdPgroup-canonical#member administrator model-mycontroller/mymodel
```

> See more: {doc}`juju add-permission <../reference/jaas-plugin>`

### Limitations when checking indirect access

IdP group membership is provided in the group claim when a user logs in. As a result, commands that need to evaluate another user's permissions cannot determine that user's indirect access through an IdP. 

For example, `juju jaas check-permission` cannot reliably check a user's access inherited from an IdP group, and `juju show-model` does not list users whose access comes through IdP group membership. The same limitation applies to any command that relies on another user's permissions.

(configure-jimm-IdP-groups)=
## Configure JIMM's charm to accept group claims
There are two ways to configure JIMM's charm for IdP group claims, and it depends on the way the Identity Platform was configured.

> If the identity platform was configured to emit the group claim by default you will need to change the `oauth-group-claim-key` charm config with the key JIMM will find the group claim in the extra claims (e.g. `groups`).

Reviewer 3: ```suggestion
If the identity platform was configured to emit the group claim by default you will need to change the `oauth-group-claim-key` charm config with the key where JIMM will find the group claim (e.g. `groups`).
```

> If the identity platoform was configured to offer the group claim as an option, in addition to previous configuration you'll need to configure `oauth-client-credential-scopes` and `oauth-optional-scopes` (e.g. `groups`) to inform the Identity Platform to emit the optional scopes for users and service accounts.

Reviewer 3: ```suggestion
If the identity platform is configured to _optionally_ offer the group claim then additionally configure both `oauth-client-credential-scopes` and `oauth-optional-scopes` (e.g. `groups`) to request group claims when authenticating client-credentials and users, respectively.
```
